interpreter/syntax/types.ml

```diff
@@ -9,26 +9,27 @@ type mut = Cons | Var
 type init = Set | Unset
 type 'a limits = {min : 'a; max : 'a option}
 
-type type_addr = ..
-type var = Stat of type_idx | Dyn of type_addr
+type var = StatX of type_idx
 
 type num_type = I32T | I64T | F32T | F64T
 type vec_type = V128T
-type heap_type = FuncHT | ExternHT | DefHT of var | BotHT
-type ref_type = null * heap_type
-type val_type = NumT of num_type | VecT of vec_type | RefT of ref_type | BotT
+type heap_type = FuncHT | ExternHT | VarHT of var | DefHT of def_type | BotHT
+and ref_type = null * heap_type
+and val_type = NumT of num_type | VecT of vec_type | RefT of ref_type | BotT
 
-type result_type = val_type list
-type instr_type = InstrT of result_type * result_type * local_idx list
-type func_type = FuncT of result_type * result_type
-type cont_type = ContT of var
-type def_type = DefFuncT of func_type | DefContT of cont_type
+and result_type = val_type list
+and instr_type = InstrT of result_type * result_type * local_idx list
+and func_type = FuncT of result_type * result_type
+and cont_type = ContVarT of var
+              | ContT of func_type
+and def_type = DefFuncT of func_type | DefContT of cont_type
 
 type table_type = TableT of Int32.t limits * ref_type
 type memory_type = MemoryT of Int32.t limits
 type global_type = GlobalT of mut * val_type
 type local_type = LocalT of init * val_type
-type tag_type = TagT of var
+type tag_type = TagVarT of var
```

Comment:
I think a tag_type should always be a TagT of func_type. The index should instead occur directly on TagImport and in the tag record, analogous to how function types are handled.

Similarly, cont_type should only be ContT of func_type, whereas DefContT carries a var.

Reply:
> Similarly, cont_type should only be ContT of func_type, whereas DefContT carries a var.

Which index space should the `var` on `DefContT` point into? Do we need a separate continuation space now?

Reply:
OK, but wouldn't that render `cont_type` obsolete then?

Reply:
Ah, I see what you mean. Yeah, we need to be able to represent cont_types after substitution, so that it can occur in types during validation.

There was a similar problem in the GC proposal with supertype declarations. What I did there is generalise to the equivalent of
```
type tag_type = TagT of heap_type
```
since heap_type subsumes both type var and inlined func_type.

Reply:
Thanks. That's what I was leaning towards. I have tried implementing it, but I seem to have a substitution error somewhere. Currently, the nested generator example fails in `cont.wast` with
```
./test/core/cont.wast:425.3-425.64: link failure: incompatible import type for "scheduler" "spawn": expected tag func [(ref cont 1)] -> [], got tag func [(ref cont 0)] -> []
```